trustedservices/local.go

```diff
@@ -0,0 +1,236 @@
+		},
+		"EndorsementStoreParams": {
+			Kind:     uint32(reflect.Map),
+			Path:     "endorsement.store_params",
+			Required: common.ParamNecessity_OPTIONAL,
+		},
+		"TrustAnchorStoreName": {
+			Kind:     uint32(reflect.String),
+			Path:     "trust_anchor.store_name",
+			Required: common.ParamNecessity_REQUIRED,
+		},
+		"TrustAnchorStoreParams": {
+			Kind:     uint32(reflect.Map),
+			Path:     "trust_anchor.store_params",
+			Required: common.ParamNecessity_OPTIONAL,
+		},
+	})
+}
+
+type LocalClientConnector struct {
+}
+
+func (c LocalClientConnector) Connect(
+	host string,
+	port int,
+	params map[string]string,
+) (common.ITrustedServicesClient, error) {
+	paramStore, err := NewLocalClientParamStore()
+	if err != nil {
+		return nil, err
+	}
+
+	err = paramStore.PopulateFromStringMapString(params)
+	if err != nil {
+		return nil, err
+	}
+
+	var client LocalClient
+
+	err = client.Init(paramStore)
+
+	return &client, err
+}
+
+type LocalClient struct {
+	PluginLocations  []string
+	Schemes          map[common.AttestationFormat]*common.SchemePlugin
+	TrustAnchorStore kvstore.KVStore
+	EndorsementStore kvstore.KVStore
+}
+
+func (c *LocalClient) Init(params *common.ParamStore) error {
+	var err error
+
+	c.PluginLocations, err = params.TryGetStringSlice("PluginLocations")
+	if err != nil {
+		return err
+	}
+
+	c.Schemes = make(map[common.AttestationFormat]*common.SchemePlugin)
+
+	storeName, err := params.TryGetString("TrustAnchorStoreName")
+	if err != nil {
+		return err
+	}
+
+	switch storeName {
+	case "memory":
+		c.TrustAnchorStore = new(kvstore.Memory)
+	case "sql":
+		c.TrustAnchorStore = new(kvstore.SQL)
+	default:
+		return fmt.Errorf("unknown TrustAnchorStoreName: %q", storeName)
+	}
+
+	storeParams, err := params.TryGetStringMap("TrustAnchorStoreParams")
+	if err != nil {
+		return err
+	}
+
+	err = c.TrustAnchorStore.Init(kvstore.Config(storeParams))
+	if err != nil {
+		return err
+	}
+
+	storeName, err = params.TryGetString("EndorsementStoreName")
+	if err != nil {
+		return err
+	}
+
+	switch storeName {
+	case "memory":
+		c.EndorsementStore = new(kvstore.Memory)
+	case "sql":
+		c.EndorsementStore = new(kvstore.SQL)
+	default:
+		return fmt.Errorf("unknown EndorsementStoreName: %q", storeName)
+	}
+
+	storeParams, err = params.TryGetStringMap("EndorsementStoreParams")
+	if err != nil {
+		return err
+	}
+
+	err = c.EndorsementStore.Init(kvstore.Config(storeParams))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *LocalClient) GetAttestation(token *common.AttestationToken) (*common.Attestation, error) {
+	scheme, err := c.getSchemePlugin(token.Format)
+	if err != nil {
+		return nil, err
+	}
+
+	ec, err := c.extractEvidence(scheme, token)
+
+	endorsementString, err := c.EndorsementStore.Get(ec.SoftwareId)
+	if err != nil {
+		return nil, err
+	}
+
+	return scheme.GetAttestation(ec, endorsementString)
+}
+
+func (c LocalClient) Close() error {
+	var msg string
```

Comment:
Just a suggestion: `msg` could be a slice of strings to which you `append()` and at the end call `strings.Join(msg, "; ")` to have it properly formatted.